src/OrleansRuntime/Services/GrainServiceClient.cs

```diff
@@ -0,0 +1,73 @@
+﻿using System;
+using Orleans.CodeGeneration;
+using Orleans.Runtime.ConsistentRing;
+using Orleans.Services;
+
+namespace Orleans.Runtime.Services
+{
+    /// <summary>
+    /// Proxies requests to the appropriate GrainService based on the appropriate Ring partitioning strategy.
+    /// </summary>
+    /// <typeparam name="TGrainService"></typeparam>
+    public abstract class GrainServiceClient<TGrainService> : IGrainServiceClient<TGrainService> where TGrainService : IGrainService
+    {
+        private readonly InsideRuntimeClient runtimeClient;
+        private readonly IInternalGrainFactory grainFactory;
+        private readonly IConsistentRingProvider ringProvider;
+        private readonly int grainTypeCode;
+
+        /// <summary>
+        /// Currently we only support a single GrainService per Silo, when multiple are supported we will request the number of GrainServices to partition per silo here.
+        /// </summary>
+        protected GrainServiceClient(IServiceProvider serviceProvider)
+        {
+            runtimeClient = (InsideRuntimeClient) serviceProvider.GetService(typeof(IRuntimeClient));
+            grainFactory = (IInternalGrainFactory) serviceProvider.GetService(typeof(IInternalGrainFactory));
+            ringProvider = (IConsistentRingProvider) serviceProvider.GetService(typeof(IConsistentRingProvider));
+
+            // GrainInterfaceMap only holds IGrain types, not ISystemTarget types, so resolved via Orleans.CodeGeneration.
+            // Resolve this before merge.
+            grainTypeCode = GrainInterfaceUtils.GetGrainClassTypeCode(typeof(TGrainService));
+        }
+
+        /// <summary>
+        /// Resolves the correct GrainService responsible for actioning the request based on the CallingGrainReference
+        /// </summary>
+        protected TGrainService GrainService
+        {
+            get
+            {
+                // Currently we only support a single GrainService per Silo, hence we pass Zero here instead of partitioning.
+                var grainId = GrainId.GetGrainServiceGrainId(0, grainTypeCode);
+
+                var destination = MapGrainReferenceToSiloRing(CallingGrainReference);
+
+                var grainService = grainFactory.GetSystemTarget<TGrainService>(grainId, destination);
+
+                return grainService;
```

Comment:
Ah, scratch that: I didn't realise that `CallingGrainReference` changes depending on who's asking for it.